flang/test/Fir/invliad-types.fir

```diff
@@ -0,0 +1,151 @@
+// Test the FIR types parser diagnostics
+// RUN: fir-opt -split-input-file -verify-diagnostics %s
+
+// expected-error@+1 {{expected non-function type}}
+func private @box3() -> !fir.boxproc<>
+
+// -----
+
+// expected-error@+2 {{expected non-function type}}
+// expected-error@+1 {{expected affine map}}
+func private @box1() -> !fir.box<!fir.array<?xf32>, >
+
+// -----
+
+// expected-error@+1 {{expected non-function type}}
+func private @box1() -> !fir.box<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @box2() -> !fir.boxchar<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @it6() -> !fir.char<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @it6() -> !fir.char<2, >
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @it3() -> !fir.complex<>
+
+// -----
+
+// expected-error@+1 {{expected non-function type}}
+func private @mem3() -> !fir.heap<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @it1() -> !fir.int<A>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @it1() -> !fir.logical<b>
+
+// -----
+
+// expected-error@+1 {{expected non-function type}}
+func private @mem3() -> !fir.ptr<>
+
+// -----
+
+// expected-error@+1 {{expected integer value}}
+func private @mem3() -> !fir.real<>
+
+// -----
+
+// expected-error@+1 {{expected valid keyword}}
+func private @mem3() -> !fir.type<>
+
+// -----
+
+// expected-error@+2 {{expected valid keyword}}
+// expected-error@+1 {{expected LEN parameter list}}
+func private @dvd4() -> !fir.type<derived4()>
+
+// -----
+
+// expected-error@+2 {{expected ':'}}
+// expected-error@+1 {{expected LEN parameter list}}
+func private @dvd4() -> !fir.type<derived4(p)>
+
+// -----
+
+// expected-error@+2 {{expected non-function type}}
+// expected-error@+1 {{expected LEN parameter list}}
+func private @dvd4() -> !fir.type<derived4(p:)>
+
+// -----
+
+// expected-error@+2 {{expected valid keyword}}
+// expected-error@+1 {{expected field type list}}
+func private @dvd4() -> !fir.type<derived4(p:i8){}>
+
+// -----
+
+// expected-error@+2 {{expected valid keyword}}
+// expected-error@+1 {{expected field type list}}
+func private @dvd4() -> !fir.type<derived4(p:i8){}>
```

Comment:
Note: we may want to allow an empty field list.

Does this need to test the fields and "listiness"? e.g. `{f1}`, `{f1:}`, `{f1:f2}`, `{f1:i32,}`, `{,}`

Reply:
I just added couple of tests for the field list.